cmd/download.go

```diff
@@ -5,166 +5,258 @@
+		KindSpec:      kindSpec,
+	}
+}
+
+// getConsumableURL returns an url that can be used for download
+func (n *PackageURL) getConsumableURL() string {
+
+	if !n.Registry {
+		return n.getURLFromCompanyRepos()
+	}
+
+	return fmt.Sprintf("%s/%s%s?ref=%s", n.KindSpec.pickCloudProviderURL(n.CloudProvider), n.Blocks[0], ".git", n.Version)
+
+}
+
+func (n *PackageURL) getURLFromCompanyRepos() string {
+	if len(n.Blocks) == 0 {
+		return ""
+	}
+
+	dG := ""
+
+	if strings.HasPrefix(n.Prefix, "git::https") {
+		dG = ".git"
+	}
+
+	if len(n.Blocks) == 1 {
+		return fmt.Sprintf("%s-%s%s//%s?ref=%s", n.Prefix, n.Blocks[0], dG, n.Kind, n.Version)
+	}
+
+	remainingBlocks := ""
+
+	for i := 1; i < len(n.Blocks); i++ {
+		remainingBlocks = path.Join(remainingBlocks, n.Blocks[i])
+	}
+
+	return fmt.Sprintf("%s-%s%s//%s/%s?ref=%s", n.Prefix, n.Blocks[0], dG, n.Kind, remainingBlocks, n.Version)
+
+}
+
+func downloadProcess(wg *sync.WaitGroup, opts DownloadOpts, data Package, errChan chan<- error, i int) {
+	// deferring the worker to be done
+	defer wg.Done()
+
+	logrus.Debugf("%d : received data %v", i, data)
+
+	// Checking git clone protocol
+	p := ""
+	if opts.Https {
+		p = httpsRepoPrefix
 	} else {
-		numberOfWorkers = 1
+		p = sshRepoPrefix
+	}
+
+	// Create the package URL from the data received to download the package
+	pU := newPackageURL(
+		p,
+		strings.Split(data.Name, "/"),
+		data.Kind,
+		data.Version,
+		data.Registry,
+		data.ProviderOpt,
+		data.ProviderKind)
+
+	u := pU.getConsumableURL()
+
+	downloadErr := get(u, data.Dir, getter.ClientModeDir, true)
+
+	// Checking if repository was found otherwise fallback to the old prefix, if fallback fails sends error to the error channel
+	if downloadErr != nil && strings.Contains(downloadErr.Error(), "Repository not found") {
+		o := humanReadableSource(pU.getConsumableURL())
+
+		if opts.Https {
+			pU.Prefix = fallbackHttpsRepoPrefix
+		} else {
+			pU.Prefix = fallbackSshRepoPrefix
+		}
+
+		logrus.Warningf("error downloading %s, falling back to %s", o, humanReadableSource(pU.getConsumableURL()))
+
+		downloadErr = get(pU.getConsumableURL(), data.Dir, getter.ClientModeDir, true)
+		if downloadErr != nil {
+			err := os.RemoveAll(data.Dir)
+			if err != nil {
+				logrus.Errorf("error removing directory %s: %s", data.Dir, err.Error())
+			}
+			errChan <- downloadErr
+		}
 	}
+}
+
+func Download(packages []Package, opts DownloadOpts) error {
+	//Preparing all the necessary data for a worker pool
+	var wg sync.WaitGroup
 	errChan := make(chan error, len(packages))
 	jobs := make(chan Package, len(packages))
-	logrus.Debugf("workers = %d", numberOfWorkers)
 
 	// Populating the job channel with all the packages to downlaod
 	for _, p := range packages {
 		jobs <- p
 	}
 
-	// Starting all the workers necessary
-	for i := 0; i < numberOfWorkers; i++ {
+	logrus.Debugf("workers = %d", len(jobs))
+
+	// Starting the workers to download the packages in parallel
+	for i, data := range packages {
```

Comment:
I find the previous approach (that used numberOfWorkers) to be preferable as:
- it allowed the user to choose between parallel and serial execution
- it took available CPUs into account instead of fanning out without control